### PR TITLE
[DO NOT MERGE] Prevmeta

### PR DESCRIPTION
**Pull request by moneromooo-monero**

### CODE REVIEW

**Comment by moneromooo-monero** [comment]
In fact, I'm wondering whether this should not be automatic.. try to open normally, then on error, try again with the flag...

**Comment by vtnerd** [comment]
Seems like @hyc should be pinged for this.

**Comment by hyc** [comment]
I have already approved and merged an updated version of the LMDB patch upstream.
We're still thinking about whether/how to automate its use in monerod.

**Comment by moneromooo-monero** [comment]
I just made it auto in that last update.

**Comment by hyc** [comment]
Not sure that change is adequate. Fatal errors are seldom detected in mdb_env_open, they usually happen later when you actually try to fetch a record.

**Comment by moneromooo-monero** [comment]
Fair point. Maybe it's best as a separate --salvage-db flag.

**Comment by hyc** [comment]
I'm using something like this now for automatic attempt
[dif.txt](https://github.com/monero-project/monero/files/1195421/dif.txt)

**Comment by moneromooo-monero** [comment]
Weird, I'm getting an amazon error page clicking on that.
Anyway, are these calls thought to be the vast majority of what could fail, but not fail using the previous meta page ? If not, a separate flag makes more sense.

**Comment by hyc** [comment]
I think it's most common for something to go wrong in these early calls, yes. But we can go with the manual flag.

**Comment by moneromooo-monero** [comment]
I think I'll just add a "you may want to retry with --db-salvage" if those error out. Gives the user some feeling of control :P

File: src/cryptonote_core/cryptonote_core.cpp [hunk 1]
```diff
@@ -307,12 +309,14 @@ namespace cryptonote
...
     uint64_t DBS_FASTEST_MODE = 0;
     uint64_t DBS_SAFE_MODE = 0;
+    uint64_t DBS_SALVAGE = 0;
     if (db_type == "lmdb")
     {
       db = new BlockchainLMDB();
       DBS_SAFE_MODE = MDB_NORDAHEAD;
       DBS_FAST_MODE = MDB_NORDAHEAD | MDB_NOSYNC;
       DBS_FASTEST_MODE = MDB_NORDAHEAD | MDB_NOSYNC | MDB_WRITEMAP | MDB_MAPASYNC;
+      DBS_SAFE_MODE = MDB_PREVMETA;
```
Comment (hyc) [inline src/cryptonote_core/cryptonote_core.cpp]: Should be DBS_SALVAGE = ...

Comment (moneromooo-monero) [inline src/cryptonote_core/cryptonote_core.cpp]: urgh

**Comment by hyc** [comment]
To avoid divergence with upstream, pls make sure you use the updated LMDB patch too. Thanks.

**Comment by moneromooo-monero** [comment]
Yes, I updated it after changing from comments.

**Comment by hyc** [comment]
After further review, the current LMDB patch isn't working as intended. This needs more thought.

**Comment by hyc** [comment]
OK, we've talked through some of the consistency concerns and addressed them here. Also renamed the flag to MDB_PREVSNAPSHOT to be a little clearer.

[0001-ITS-8704-add-MDB_PREVSNAPSHOT-flag-to-mdb_env_open.patch.txt](https://github.com/monero-project/monero/files/1219710/0001-ITS-8704-add-MDB_PREVSNAPSHOT-flag-to-mdb_env_open.patch.txt)

**Pull request closed without merging**